ops/_private/harness.py

```diff
@@ -161,18 +172,35 @@ class ActionOutput:
     """The action's results, as set or updated by :meth:`ops.ActionEvent.set_results`."""
 
 
-class ActionFailed(Exception):  # noqa
-    """Raised when :code:`event.fail()` is called during a :meth:`Harness.run_action` call."""
+class ActionFailed(Exception):  # noqa: N818 (name doesn't end with "Error")
+    """Raised when :code:`event.fail()` is called during an action handler."""
 
     message: str
     """Optional details of the failure, as provided by :meth:`ops.ActionEvent.fail`."""
 
     output: ActionOutput
-    """Any logs and results set by the Charm."""
+    """Any logs and results set by the Charm.
+
+    When using Context.run, both logs and results will be empty - these
+    can be found in Context.action_logs and Context.action_results.
+    """
+
+    state: typing.Optional['State']
+    """The Juju state after the action has been run.
 
-    def __init__(self, message: str, output: ActionOutput):
+    When using Harness.run_action, this will be None.
+    """
+
+    def __init__(
+        self,
+        message: str,
+        output: typing.Optional[ActionOutput] = None,
+        *,
+        state: typing.Optional['State'] = None,
+    ):
         self.message = message
-        self.output = output
+        self.output = ActionOutput([], {}) if output is None else output
```

Comment:
This would be the same, wouldn't it?

```suggestion
        self.output = output or ActionOutput([], {})
```